Python code formatting standard, automatic.

dannon opened the pull request:
Automatic formatting via `black` and import sorting with `isort` of the galaxy codebase.  To run the full thing locally, make sure you have python dev dependencies installed and then just run `make format`.

I tried to do this a few years ago, but recent PRs (including the one this builds on, #11155) make me hopeful that we're ready for it now.

~WIP, draft, I need to automate it and add github actions and actual dependencies, but~ I wanted to open this for discussion before putting in the extra legwork if this is going to be DOA.  I really hope it is not, though.  Prettier, even when it does weird stuff sometimes, has been a huge boon in having all the javascript code just look the same regardless of author, with no arguments or one person formatting one way, another formatting back a different way, etc.

Script example showing how to rebase pre-black branches easily here:  https://gist.github.com/ndevenish/a7a9eb93cc19abbfe80f2daab16d3367.  Note this uses 'master', and you'd likely be targeting 'dev', though.

dannon committed:
Introduce pyproject.toml for black settings

mvdbeek committed:
Configure isort to produce the right import order style

mvdbeek committed:
Move conditional import to end of section

This is necessary so that isort doesn't add extra lines between
statements and import order linting doesn't fail.

mvdbeek committed:
Drop python3.6 compat import

mvdbeek committed:
Increase line length so long import statements aren't split

mvdbeek committed:
Move config to .isort.cfg

dannon committed:
Add black to dev-dependencies

dannon committed:
Add black check action to python linting

dannon committed:
Add '.git-blame-ignore-revs' file for better git blame usage with
sweeping changes.  To use this locally just do:
`git config blame.ignoreRevsFile .git-blame-ignore-revs`
And your git blame will now skip revisions indicated in the file.

Now that we have gitignore-ignore-revs, we might as well use it for the big prettier passes

Add isort application to .git-blame-ignore-revs

dannon committed:
Update shared pyfaidx to match (regenerated from poetry, this is the only change that should be required)

dannon committed:
E203 handling -- see discussion in See https://black.readthedocs.io/en/stable/faq.html#why-are-flake8-s-e203-and-w503-violated

nsoranzo commented on pyproject.toml:
```diff
@@ -91,6 +91,7 @@ zipstream-new = "*"
 
 [tool.poetry.dev-dependencies]
 cwltest = "2.2.20210901154959"
+black = "^22.1.0"
```
```suggestion
black = "^22.1.0"
cwltest = "2.2.20210901154959"
```

nsoranzo commented on pyproject.toml:
```diff
@@ -120,3 +121,14 @@ testfixtures = "*"
 twill = "*"
 watchdog = "*"
+
+[tool.black]
```
I'd move this section to the top to keep the sections ordered alphabetically.

dannon replied:
Will do!

nsoranzo commented on setup.cfg:
```diff
@@ -7,9 +7,10 @@ universal = 1
 # W503 is line breaks before binary operators, which has been reversed in PEP 8.
 # D** are docstring linting - which we mostly ignore except D302. (Hopefully we will solve more over time).
-ignore = B008,E128,E501,E402,W503,D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D204,D205,D206,D207,D208,D209,D210,D211,D300,D301,D400,D401,D402,D403,D412,D413
+ignore = B008,E128,E203,E501,E402,W503,D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D204,D205,D206,D207,D208,D209,D210,D211,D300,D301,D400,D401,D402,D403,D412,D413
```
```suggestion
ignore = B008,E128,E203,E402,E501,W503,D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D204,D205,D206,D207,D208,D209,D210,D211,D300,D301,D400,D401,D402,D403,D412,D413
```

nsoranzo commented on setup.cfg:
```diff
@@ -7,9 +7,10 @@ universal = 1
 # E128 continuation line under-indented for visual indent
 # E402 module level import not at top of file # TODO, we would like to improve this.
 # E501 is line length
+# E203 is whitespace before ':'; we follow black's formatting here.  See https://github.com/psf/black/issues/315
```
```suggestion
# E203 is whitespace before ':'; we follow black's formatting here. See https://black.readthedocs.io/en/stable/faq.html#why-are-flake8-s-e203-and-w503-violated
# E402 module level import not at top of file # TODO, we would like to improve this.
# E501 is line length
```

dannon committed:
Black-related mypy fixes

nsoranzo commented on pyproject.toml:
```diff
@@ -120,3 +121,14 @@ testfixtures = "*"
+target-version = ['py37']
+include = '\.pyi?$'
+extend-exclude = '''
+/(
```
```suggestion
^/(
```

Otherwise directories like `lib/galaxy/tools` are excluded.

nsoranzo commented on pyproject.toml:
```diff
@@ -120,3 +121,14 @@ testfixtures = "*"
+  | packages
+  | tools
+)/
+'''
```
```suggestion
'''
force-exclude = 'lib/galaxy/util/jstree.py'
```

We also ignore this file for flake8 in `setup.cfg` . Changes to `lib/galaxy/util/jstree.py` should also be reverted, I think.

nsoranzo reviewed (commented):
Thanks @dannon ! What about integrating the changes from #13270 in here as well so we have a single commit to ignore that does both isort and black?

dannon committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

dannon committed:
Reorder pyproject.toml

dannon committed:
Apply black formatting.

Apply isort.

dannon committed:
Add isort devdependency

dannon committed:
Isort config update from Nicola at https://github.com/galaxyproject/galaxy/pull/13270\#discussion_r798123586

dannon committed:
Fix comments in .isort.cfg -- looks like you can't have inline comments here?

dannon committed:
Drop flake8-import-order (this will all be automagically reordered by isort, so there's no need to lint the order)

dannon committed:
Ignore E203 even for stricter linting subset

dannon committed:
Add simple make target for formatting everything (non-diff)

nsoranzo commented on Makefile:
```diff
@@ -41,6 +41,10 @@ setup-venv:
 
+format:
+	$(IN_VENV) black .
+	$(IN_VENV) isort .
```
```suggestion
	$(IN_VENV) isort .
	$(IN_VENV) black .
```

I feel black should have the last word?

dannon replied:
Sure, good idea.  I did test running them repeatedly in different orders and they're currently not in conflict (i.e. -- after they've both been run, there are no changes when running either, so order shouldn't currently matter), but it does feel like black should go last.

nsoranzo commented on pyproject.toml:
```diff
@@ -88,8 +100,10 @@ uvloop = "*"
 WebOb = "*"
 Whoosh = "*"
 zipstream-new = "*"
+isort = "^5.10.1"
```
Shouldn't `isort` go to the `[tool.poetry.dev-dependencies]` section instead?

dannon replied:
Good catch!  Wil fix and regenerate lock and dependencies.txts to make sure they're correct.

dannon committed:
Update Makefile

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

dannon committed:
Move isort back into devDependencies.

mvdbeek commented on .git-blame-ignore-revs:
```diff
@@ -1,5 +1,7 @@
 # Migrate code style to Black
 7ac461a3191fc56909da8ac3c8294f4dd4e0d543
+# Apply isort
+11207757b81157c13070c491ffa60822dcd33a57
```
Maybe it'll be easier to rebase if this is one commit ? So once you hit the conflict, you checkout `ours` and then run both black and isort, otherwise you have to first check the conflicting commit to pick the right tool ?

dannon replied:
Sure, I'll rebase those together and update the hash

mvdbeek replied:
We should probably also have the big commit be after we setup the black and isort config, so that you already have our black and isort config in place when you hit the conflict ?

mvdbeek replied:
(which I guess is the case anyway, otherwise it's hard to run the tools anyway. nevermind)

dannon replied:
Yep, it'll be the big commit and then the hash update and that's the end

nsoranzo commented on .isort.cfg:
```diff
@@ -0,0 +1,14 @@
+reverse_relative=true
+skip_gitignore=true
+# Make isort run faster by skipping database
+skip_glob=database/*  
```
```suggestion
skip_glob=database/*
```

nsoranzo commented on pyproject.toml:
```diff
@@ -1,3 +1,15 @@
+extend-exclude = '''
+^/(
+  | packages
+  | tools
```
```suggestion
```

dannon replied:
@nsoranzo I was actually going to ask about this -- I left this out thinking that even though Galaxy is 3.7+ some of the tools might be deployed using older pythons.  If that's not a concern, then I'm happy to drop that exclude.

nsoranzo replied:
Mmmh, true, it may be that black changes the code in a 3.7+-only way.

dannon replied:
So that's exactly what I was worried about, but it occurs to me I could apply black to tools with a 3.6 profile, run again with 3.7 and see if there are any changes.  Sound good?  Then we can maybe drop the exclude.

nsoranzo replied:
There are many tools that don't specify any `python` requirement at all (which is clearly not good), and we could be faced with similar worries once we move to the next version of Python, I fear.

dannon replied:
Sorry, I mean use black with the 3.6 profile (which I assume all tools are compatible with?), not the tool profile itself.

nsoranzo committed:
Help black with placement of noqa comments

nsoranzo committed:
Format files in `packages/` with black

with:

```
find packages/ -name '*.py' | xargs black
```

The `packages/` directory is currently excluded in the `[tool.black]`
section of `pyproject.toml` because black follows symlinks.

nsoranzo commented on setup.cfg:
```diff
@@ -5,16 +5,13 @@ universal = 1
 # These are exceptions allowed by Galaxy style guidelines:
 # B008 Do not perform function calls in argument defaults (for FastAPI Depends and Body)
 # E128 continuation line under-indented for visual indent
```
```suggestion
```

nsoranzo commented on setup.cfg:
```diff
@@ -5,16 +5,13 @@ universal = 1
 # W503 is line breaks before binary operators, which has been reversed in PEP 8.
 # D** are docstring linting - which we mostly ignore except D302. (Hopefully we will solve more over time).
-ignore = B008,E128,E501,E402,W503,D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D204,D205,D206,D207,D208,D209,D210,D211,D300,D301,D400,D401,D402,D403,D412,D413
+ignore = B008,E128,E203,E402,E501,W503,D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D204,D205,D206,D207,D208,D209,D210,D211,D300,D301,D400,D401,D402,D403,D412,D413
```
```suggestion
ignore = B008,E203,E402,E501,W503,D100,D101,D102,D103,D104,D105,D106,D107,D200,D201,D202,D204,D205,D206,D207,D208,D209,D210,D211,D300,D301,D400,D401,D402,D403,D412,D413
```

nsoranzo commented on .github/workflows/lint.yaml:
```diff
@@ -43,3 +43,4 @@ jobs:
         run: tox -e lint_docstring_include_list
       - name: Run mypy checks
         run: tox -e mypy
+      - uses: psf/black@stable
```
```suggestion
      - uses: psf/black@stable
      - uses: jamescurtin/isort-action@master
```

dannon committed:
Add isort github action to python linting workflow.

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo commented on Makefile:
```diff
@@ -41,6 +41,10 @@ setup-venv:
 diff-format:
```
```suggestion
diff-format:  ## Format Python code changes
```

To add this to the help print by `make` with no targe.

nsoranzo commented on Makefile:
```diff
@@ -41,6 +41,10 @@ setup-venv:
 diff-format:
 	$(IN_VENV) darker -r $(TARGET_BRANCH)
 
+format:
```
```suggestion
format:  ## Format Python code base
```

dannon committed:
Apply suggestions from code review

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

nsoranzo committed:
Update Python formatting docs

dannon committed:
Apply black to tools as well

dannon committed:
Add /tools/ application of black to git-blame-ignore-revs as well.  Not rebasing this one into the main blob and rerunning because I don't expect merge conflicts like we'd have in the core codebase.

jmchilton reviewed (approved):
Much easier to approve with the trailing commas and with Marius and Nicola taking over the CWL branch lately. Thanks for all the work on this Dannon - really nice job.

mvdbeek reviewed (approved):
Awesome, thanks @dannon & @nsoranzo
